### PR TITLE
fix(#3306): useComboBox: Escape key should continue propagation, when appropriate, to close parent Overlay

### LINKED ISSUE #3306: useComboBox: Escape key should continue propagation, when appropriate, to close parent Overlay
# 🐛 Bug Report 

With focus on a ComboBox input that appears within an Overlay, pressing the Escape key should continuePropagation to close the Overlay when:
1. a selectedKey has been set,
2. the inputValue is empty,
3. or the inputValue is not empty and `allowsCustomValue` is true.

## 🤔 Expected Behavior

It should be possible to close an Overlay by pressing the Escape key with focus on a ComboBox within the Overlay.

## 😯 Current Behavior

Escape key will either close the Listbox overlay for the ComboBox if it is open, or clear the typed value, but once a value has been set with a `selectedKey` when `allowsCustomValue` is true, the keydown event does not propagate as expected to close the Overlay. A user has to tab away from the ComboBox for the Escape key to work.

## 💁 Possible Solution

At https://github.com/adobe/react-spectrum/blob/f0c9a597aa8d9eadeb47e81aacd9af65dca77700/packages/%40react-aria/combobox/src/useComboBox.ts#L121-L123, add the following:
```ts
      case 'Escape':
        if (
          state.selectedKey !== null ||
          state.inputValue === '' ||
          props.allowsCustomValue
        ) {
          e.continuePropagation();
        }
        state.revert();
        break;
```

## 🔦 Context

Issue identified in Admin Console: ONESIE-29933

## 💻 Code Sample

A possible workaround adding an additional onKeyDown event handler to the ComboBox instance within a Dialog: https://codesandbox.io/s/ecstatic-dream-qqesyl?file=/src/ComboBoxInDialogWorkaround.js

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   | @react-aria/combobox@3.3.1
| Browser          | Chrome Version 103.0.5060.114 (Official Build) (arm64)
| Operating System | macOS 12.4 (21F79)

## 🧢 Your Company/Team

Adobe/Accessibility

## 🕷 Tracking Issue (optional)

ONESIE-29933

### PR DESCRIPTION
**Pull request by majornista**
Closes #3306 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding React Spectrum GitHub Issue #3306.
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

1. open https://reactspectrum.blob.core.windows.net/reactspectrum/821a9e31789bb3021d33ba0e97fcb532669f75a9/storybook/index.html?path=/story/combobox--within-a-dialog&providerSwitcher-toastPosition=bottom
2. Tab to set focus to the DialogTrigger Button
3. Press Space or Enter to open the Dialog
4. Tab to set focus to the ComboBox within the Dialog
5. Enter some arbitrary text into the ComboBox input
6. Press Esc key to clear the input value.
7. Press Esc key again with the input value empty.
8. Dialog should close, restoring focus to the DialogTrigger Button
9. Press Space or Enter to open the Dialog
10. Tab to set focus to the ComboBox within the Dialog
11. Press ArrowDown to expand the ComboBox dropdown listbox.
12. Use Arrow keys to navigate items in the listbox
13. Select an item by pressing Enter.
14. Listbox dropdown should close, restoring focus to the input containing the selected value.
15. Press Esc key, which with a selected key should close the Dialog, restoring focus to the DialogTrigger Button.

16. https://reactspectrum.blob.core.windows.net/reactspectrum/821a9e31789bb3021d33ba0e97fcb532669f75a9/storybook/index.html?path=/story/combobox--within-a-dialog-allowscustomvalue-true&providerSwitcher-toastPosition=bottom
17. Tab to set focus to the DialogTrigger Button
18. Press Space or Enter to open the Dialog
19. Tab to set focus to the ComboBox within the Dialog
20. Enter some arbitrary text into the ComboBox input
21. Press Esc key.
22. The Dialog should close, because with `allowsCustomValue`, any value is acceptable within the ComboBox, so Esc key event should `continuePropagation`.

## 🧢 Your Project:

Adobe/Accessibility

### CODE REVIEW

**Review by snowystinger — APPROVED** [review]
Makes sense to me

**Review by snowystinger — APPROVED** [review]
have a re-approval

File: packages/@react-aria/combobox/src/useComboBox.ts [hunk 1]
```diff
@@ -119,6 +119,13 @@ export function useComboBox<T>(props: AriaComboBoxOptions<T>, state: ComboBoxSta
         state.commit();
         break;
       case 'Escape':
+        if (
+          state.selectedKey !== null ||
+          state.inputValue === '' ||
+          props.allowsCustomValue
```
Comment (devongovett) [inline packages/@react-aria/combobox/src/useComboBox.ts:125]: One weird case here is if you set a custom value, and then press escape, it both resets the combobox to its previous selected value, and also closes the dialog. Probably should only reset first, and then a second press should close it.

Also initially looking at the logic here it seemed like there should be some condition about the menu also being closed, but somehow that is already happening. Do you know why?

**Review by devongovett — APPROVED** [review]
Approving because it's much better than before. We can follow up on the one potential bug I found above.

**Pull request merged**